Adding configurable Docker registry

Graeme-Miller opened the pull request:
Updated version of this [pull request](https://github.com/brooklyncentral/clocker/pull/168) with a slightly better branch name.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/DockerContainerImpl.java:
```diff
@@ -28,6 +28,7 @@
 
 import javax.annotation.Nullable;
 
+import com.google.common.collect.*;
```
Don't use wildcards for imports - should be configurable in IDE?

Graeme-Miller replied:
hmm, must have made this commit before I changed the IDE settings. Will go through all classes and check when we are ready to merge

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/DockerHostImpl.java:
```diff
@@ -692,6 +701,31 @@ public void postStart() {
+                    throw new IllegalArgumentException(String.format("registry {%s} and username {%s} supplied but password in DOCKER_EXTERNAL_REGISTRY_LOGIN_DETAILS is null", registry, username));
+                }
+
+                runDockerCommand(String.format("login  -e \"fake\" -u %s -p %s %s", username, password, registry));
```
Should we maybe make email configurable?

Graeme-Miller replied:
Maybe. Trouble is, we don't really need it here. Email is *usually* only used by the login command when registering a new user. When we use login here, we only really want to login, not register. 

The annoying thing is you still have to provide a non null email address. I suppose I could create an email config option with "fake" as the default?

There is an open git issue about this for docker [here](https://github.com/docker/docker/issues/6400)

grkvlt replied:
Meh. Change it to `fake@example.org` then, to make it clear it's an email address, and an unusable one, at thet.

grkvlt commented:
Should squash these into a single commit (or a few if it makes more sense) before merging. The formatting also needs updated, and can be done at the same time.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java:
```diff
@@ -427,6 +421,32 @@ public void stop() {
+
+            LOG.debug("Starting a new Docker Registry with spec {}", spec);
+            dockerRegistry.start(ImmutableList.of(firstEntity.getDynamicLocation()));
+
```
Better to use `Entities#Start(Coillection)` for this, which will manage the entity and call start in a new `Task`.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistry.java:
```diff
@@ -0,0 +1,48 @@
+                            "REGISTRY_HTTP_TLS_KEY", "/certs/repo-key.pem"
+            ));
+
+    @SetFromFlag("dockerHost")
```
Not useful on `AttributeSensor` fields

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistry.java:
```diff
@@ -0,0 +1,48 @@
+            ));
+
+    @SetFromFlag("dockerHost")
+    AttributeSensor<String[]> DOCKER_REGISTRY_CATALOG = Sensors.newSensor(new TypeToken<String[]>(){}, "docker.registry.catalog", "The docker registry catalog, which lists all of the available repositories");
```
Prefer typed `List<String>` to arrays

Graeme-Miller replied:
Me too, but it wouldn't work with List<String>
The JSON parser that populates this needs an array rather than a List

grkvlt replied:
You just need to iterate through the JSON objects. I'll paste an example at the relevant comment.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistryImpl.java:
```diff
@@ -0,0 +1,109 @@
+
+public class DockerRegistryImpl extends VanillaDockerApplicationImpl implements DockerRegistry {
+
+    private volatile HttpFeed httpFeed;
```
Should be `transient` to indicate ignore during serialisation. The `volatile` modifier is for concurrency. See this answer on [stackoverflow](http://stackoverflow.com/questions/3544919/what-are-transient-and-volatile-modifiers) for a better description.

Graeme-Miller replied:
I marked this as  volatile as I was following the precedent set in other classes, for example, RiakNodeImpl. Can remove if you think there will be no multi-threading issues?

Not sure about adding transient, is that something you would recommend? When would this be serialised? Possibly as part of persistence?

grkvlt replied:
Hm, probably cargo-cult copying going on there. There will never be multiple threads trying to read the reference to the feed. But, during persistance it may be serialized, and it has anonymous classes inside, so we don't want to do that. Really, `transient` is right, `volatile` is wrong and should be fixed in brooklyn - i'll review the code for that...

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistryImpl.java:
```diff
@@ -0,0 +1,109 @@
+        super.connectSensors();
+        connectServiceUpIsRunning();
+
+        HostAndPort hostAndPort = BrooklynAccessUtils.getBrooklynAccessibleAddress(this, this.getConfig(DockerRegistry.DOCKER_REGISTRY_PORT));
```
Use `config().get(...)` instead of `this.getConfig(...)`

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistryImpl.java:
```diff
@@ -0,0 +1,109 @@
+                        .onFailureOrException(Functions.constant(false)))
+                .poll(new HttpPollConfig<String[]>(DockerRegistry.DOCKER_REGISTRY_CATALOG)
+                        .suburl("/_catalog")
+                        .onSuccess(HttpValueFunctions.jsonContents("repositories", String[].class)))
```
This will need updated when the attribute type is changed to a `List`

grkvlt replied:
This example should work?

```Java
                .poll(new HttpPollConfig<List<String>>(DOCKER_REGISTRY_CATALOG)
                        .suburl("/_catalog")
                        .onSuccess(Functionals.chain(
                                HttpValueFunctions.jsonContents(),
                                JsonFunctions.walk("repositories"),
                                JsonFunctions.forEach(JsonFunctions.cast(String.class)))
                        .onFailureOrException(Functions.constant(Arrays.asList(new String[0]))))
```

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistryImpl.java:
```diff
@@ -0,0 +1,109 @@
+
+
+        SshMachineLocation sshMachine = host.getDynamicLocation().getMachine();
+        sshMachine.installTo("classpath://brooklyn/entity/container/docker/registry/docker-registry-create-certs.sh", Os.mergePaths(getSSHMachineInstallDir(), "docker-registry-create-certs.sh"));
```
Shove `getSSHMachineInstallDir()` in a local variable (`dockerInstallDir`) and make `docker-registry-create-certs.sh` a constant as well

grkvlt commented on docker/src/main/java/brooklyn/location/docker/DockerHostLocation.java:
```diff
@@ -159,8 +161,11 @@ public DockerContainerLocation obtain(Map<?,?> flags) throws NoMachinesAvailable
 
             Optional<String> baseImage = Optional.fromNullable(entity.config().get(DockerAttributes.DOCKER_IMAGE_NAME));
             String imageTag = Optional.fromNullable(entity.config().get(DockerAttributes.DOCKER_IMAGE_TAG)).or("latest");
+            Optional<String> imageRepo = Optional
```
Don't split a line after the classname

grkvlt commented on docker/src/main/java/brooklyn/location/docker/DockerHostLocation.java:
```diff
@@ -177,13 +182,23 @@ public DockerContainerLocation obtain(Map<?,?> flags) throws NoMachinesAvailable
+                    repoAndName = imageRepo.get()+"/";
+                }
+                fullyQualifiedName += baseImage.get() + ":" +imageTag;
+                repoAndName +=  baseImage.get();
```
```Java
String repoAndName = Joiner.on('/').join(Optional.presentInstances(ImmutableList.of(imageRepo, baseImage)));
String fullyQualifiedName = repoAndName + ":" + imageTag;
```
:frog:

grkvlt commented on patches/src/main/java/org/jclouds/docker/compute/options/DockerTemplateOptions.java:
```diff
@@ -24,6 +24,7 @@
 
 import org.jclouds.compute.ComputeService;
 import org.jclouds.compute.options.TemplateOptions;
+import org.jclouds.docker.internal.NullSafeCopies;
```
Definitely keep these files (the `jclouds-labs` PR in `patches`) as a separate commit

Graeme-Miller replied:
done

grkvlt commented:
We need to add a catalog item to `default.catalog.bom` and `clocker.bom` for the registry and update or add example blueprints with the registry configured for both calico and weave. It would also be good to have an example that used the public `clockercentral/clocker` repository at [quay.io](https://quay.io/repository/clockercentral/clocker)

grkvlt commented:
@Graeme-Miller Looks good so far. Make sure the comments are addressed and some examples created. The commits need to be rebased against `master` and then squashed and formatted as well. I think it just needs a bit more testing and we should be good to merge.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java:
```diff
@@ -23,15 +23,15 @@
 
+import brooklyn.entity.container.docker.registry.DockerRegistry;
+import org.apache.brooklyn.core.entity.AbstractApplication;
+import org.apache.brooklyn.entity.group.*;
```
No wildcard imports.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java:
```diff
@@ -23,15 +23,15 @@
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
No wildcard imports.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructureImpl.java:
```diff
@@ -427,6 +422,31 @@ public void stop() {
+        super.postStart(locations);
+
+
+        if(config().get(DOCKER_SHOULD_START_REGISTRY)){
```
Formatting, including squashing multiple blank lines to single.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistryImpl.java:
```diff
@@ -0,0 +1,110 @@
+                                HttpValueFunctions.jsonContents(),
+                                JsonFunctions.walk("repositories"),
+                                JsonFunctions.forEach(JsonFunctions.cast(String.class))))
+                        .onFailureOrException(Functions.constant(Arrays.asList(new String[0]))))
```
Maybe `Lists.<String>newArrayList()` instead of creating the empty `String[0]` array?

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistryImpl.java:
```diff
@@ -0,0 +1,110 @@
+
+        SshMachineLocation sshMachine = host.getDynamicLocation().getMachine();
+        String sshMachineInstallDir = getSSHMachineInstallDir();
+        sshMachine.installTo("classpath://brooklyn/entity/container/docker/registry/docker-registry-create-certs.sh", Os.mergePaths(sshMachineInstallDir, CREATE_CERTS_SCRIPT_NAME));
```
Put in a `public static final` constant at top of class

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/repository/DockerRepositoryDriver.java:
```diff
@@ -0,0 +1,6 @@
+
+import brooklyn.entity.container.docker.application.VanillaDockerApplicationDriver;
+
+public interface DockerRepositoryDriver extends VanillaDockerApplicationDriver {
```
Can we actually just reference the `VanillaDockerApplicationDriver` from our `DockerRepository` entity directly, without the need to create these subclasses?

Graeme-Miller commented:
Andrew, I have fixed everything as requested and added to the BOM files.
Can you please advise on those changes?
Also, you have requested examples written. Where do I put these examples?

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistryImpl.java:
```diff
@@ -0,0 +1,110 @@
+
+        SshMachineLocation sshMachine = host.getDynamicLocation().getMachine();
+        String sshMachineInstallDir = getSSHMachineInstallDir();
+        sshMachine.installTo(SCRIPT_LOCATION, Os.mergePaths(sshMachineInstallDir, CREATE_CERTS_SCRIPT_NAME));
```
```Java
installTo(MutableMap.of(SshTool.PROP_PERMISSIONS, "0755"), SCRIPT_LOCATION, Os.mergePaths(sshMachineInstallDir, CREATE_CERTS_SCRIPT_NAME));
```

This should let you just call the script with a single command below.

Graeme-Miller replied:
agreed that it was not worth re-writing at this stage

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistryImpl.java:
```diff
@@ -0,0 +1,110 @@
+    private String getSSHMachineInstallDir() {
+        return config().get(DockerRegistry.DOCKER_HOST).sensors().get(SoftwareProcess.INSTALL_DIR);
+    }
+
```
Add a `getIconUrl()` method that returns `classpath://docker-registry-logo.png`

Graeme-Miller replied:
done

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistry.java:
```diff
@@ -0,0 +1,46 @@
+import brooklyn.entity.container.docker.application.VanillaDockerApplication;
+
+
+@ImplementedBy(DockerRegistryImpl.class)
```
Add the `@Catalog` annotations here

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/DockerInfrastructure.java:
```diff
@@ -133,6 +133,11 @@
             new TypeToken<Map<String, Object>>() { },
             "docker.dockerfile.substitutions", "Dockerfile template substitutions", MutableMap.<String, Object>of());
 
+    @SetFromFlag("shouldStartRegistry")
```
Make this a catalog entry property, so we can see it when we start things from the catalog.

grkvlt commented on clocker.bom:
```diff
@@ -49,6 +49,18 @@ brooklyn.catalog:
           $brooklyn:entitySpec:
             type: docker-host
 
+  - id: docker-infrastructure-with-registry
```
Should have a `docker-registry` catalog entry` and then add the config to a new template for a `docker-cloud-with-registry` as an example.

grkvlt commented on dist/src/main/assembly/conf/brooklyn/default.catalog.bom:
```diff
@@ -193,26 +205,6 @@ brooklyn.catalog:
                     brooklyn.config:
                       docker.container.nameFormat: "weave-%2$02x"
 
-  - id: image-microservice
```
Why are these deleted? Should probably be replaced.

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistry.java:
```diff
@@ -0,0 +1,46 @@
+@ImplementedBy(DockerRegistryImpl.class)
+public interface DockerRegistry extends VanillaDockerApplication {
+
+    @SetFromFlag("imageName")
```
Remove the `@SetFromFlag` for config that should not be overwritten

grkvlt commented on docker/src/main/java/brooklyn/entity/container/docker/registry/DockerRegistry.java:
```diff
@@ -0,0 +1,46 @@
+    @SetFromFlag("imageTag")
+    ConfigKey<String> IMAGE_TAG = ConfigKeys.newConfigKeyWithDefault(DockerAttributes.DOCKER_IMAGE_TAG.getConfigKey(), "2");
+
+    AttributeSensorAndConfigKey<Integer, Integer> DOCKER_REGISTRY_PORT = DockerHost.DOCKER_REGISTRY_PORT;
```
Add `@SetFromFlag` and `@CatalogConfig` here

grkvlt commented:
Thanks @Graeme-Miller this looks good to merge into master
